// express/index.js
const http = require('http');
const { URL } = require('url');

function enhanceResponse(res) {
  res.status = function status(code) {
    res.statusCode = code;
    return res;
  };

  res.json = function json(data) {
    if (!res.headersSent) {
      res.setHeader('Content-Type', 'application/json');
    }
    res.end(JSON.stringify(data));
  };

  return res;
}

function matchRoute(pattern, actualPath) {
  if (pattern === actualPath) {
    return {};
  }

  const patternParts = pattern.split('/').filter(Boolean);
  const pathParts = actualPath.split('/').filter(Boolean);

  if (patternParts.length !== pathParts.length) {
    return null;
  }

  const params = {};

  for (let i = 0; i < patternParts.length; i += 1) {
    const part = patternParts[i];
    const value = pathParts[i];

    if (part.startsWith(':')) {
      params[part.slice(1)] = decodeURIComponent(value);
    } else if (part !== value) {
      return null;
    }
  }

  return params;
}

function express() {
  const middlewares = [];
  const routes = [];

  const app = function app(req, res) {
    enhanceResponse(res);

    const parsedUrl = new URL(req.url, `http://${req.headers.host || 'localhost'}`);
    req.path = parsedUrl.pathname;
    req.query = Object.fromEntries(parsedUrl.searchParams.entries());
    req.params = {};
    req.body = req.body || {};

    let idx = 0;

    const runMiddleware = () => {
      if (res.writableEnded) {
        return;
      }

      if (idx >= middlewares.length) {
        handleRoute();
        return;
      }

      const middleware = middlewares[idx];
      idx += 1;
      middleware(req, res, () => {
        runMiddleware();
      });
    };

    const handleRoute = () => {
      for (const route of routes) {
        if (route.method !== req.method) {
          continue;
        }
        const params = matchRoute(route.path, req.path);
        if (params) {
          req.params = params;
          route.handler(req, res);
          return;
        }
      }

      if (!res.writableEnded) {
        res.statusCode = 404;
        res.end('Not Found');
      }
    };

    runMiddleware();
  };

  app.use = (fn) => {
    middlewares.push(fn);
  };

  ['get', 'post', 'put', 'patch', 'delete', 'options'].forEach((method) => {
    app[method] = (path, handler) => {
      routes.push({ method: method.toUpperCase(), path, handler });
    };
  });

  app.listen = (port, callback) => {
    const server = http.createServer(app);
    return server.listen(port, callback);
  };

  app.locals = {};

  return app;
}

function json() {
  return (req, res, next) => {
    const contentType = req.headers['content-type'] || '';
    if (!contentType.includes('application/json')) {
      req.body = req.body || {};
      next();
      return;
    }

    let data = '';
    req.on('data', (chunk) => {
      data += chunk;
      if (data.length > 1e6) {
        req.connection.destroy();
      }
    });

    req.on('end', () => {
      if (!data) {
        req.body = {};
        next();
        return;
      }

      try {
        req.body = JSON.parse(data);
        next();
      } catch (error) {
        if (!res.writableEnded) {
          res.status(400).json({ error: 'Invalid JSON' });
        }
      }
    });

    req.on('error', () => {
      if (!res.writableEnded) {
        res.status(400).end();
      }
    });
  };
}

express.json = json;

module.exports = express;
